File: SRP/Delphi/aplicado/ProcessadorPagamentoService.pas
unit ProcessadorPagamentoService;

interface

type
  TProcessadorPagamentoService = class
    public
      procedure Processar(detalhesPagamento: TDetalhesPagamento; carrinho: TCarrinho);
  end;

implementation

procedure TProcessadorPagamentoService.Processar(detalhesPagamento: TDetalhesPagamento; carrinho: TCarrinho);
var
	gatewayPagamento: TGatewayPagamento;
begin
	gatewayPagamento := TGatewayPagamento.Create;
	
	gatewayPagamento.Credencial := "bla";
	gatewayPagamento.NumeroCartao := detalhesDoPagamento.NumeroCartao;
	gatewayPagamento.MesExpiracao := detalhesDoPagamento.MesExpiracao;
	gatewayPagamento.AnoExpiracao := detalhesDoPagamento.AnoExpiracao;
	gatewayPagamento.NomeNoCartao := detalhesDoPagamento.NomeNoCartao;
	gatewayPagamento.Valor := carrinho.ValorTotal;

	gatewayPagamento.Cobrar;
end

end.

File: DIP/C#/aplicado/ProcessadorPagamentoService.cs
interface IProcessadorPagamentoService
{
	void Processar(DetalhesPagamento detalhesDoPagamento, Carrinho carrinho);	
}

class ProcessadorPagamentoService
{
	public void Processar(DetalhesPagamento detalhesDoPagamento, Carrinho carrinho)
	{
		var gatewayPagamento = new GatewayPagamento();
		
		gatewayPagamento.credencial = "bla";
		gatewayPagamento.MumeroCartao = detalhesDoPagamento.MumeroCartao;
		gatewayPagamento.MesExpiracao = detalhesDoPagamento.MesExpiracao;
		gatewayPagamento.AnoExpiracao = detalhesDoPagamento.AnoExpiracao;
		gatewayPagamento.NomeNoCartao = detalhesDoPagamento.NomeNoCartao;
		gatewayPagamento.Valor = carrinho.ValorTotal;

		gatewayPagamento.Cobrar();		
	}
}